// release-notes/lib/release-notes-view.js
(function() {
  var ReleaseNotesView, View, shell,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  shell = require('shell');

  View = require('atom').View;

  module.exports = ReleaseNotesView = (function(_super) {
    __extends(ReleaseNotesView, _super);

    function ReleaseNotesView() {
      return ReleaseNotesView.__super__.constructor.apply(this, arguments);
    }

    ReleaseNotesView.content = function() {
      return this.div({
        "class": 'release-notes padded pane-item native-key-bindings',
        tabindex: -1
      }, (function(_this) {
        return function() {
          _this.h1({
            "class": 'section-heading',
            outlet: 'version'
          });
          _this.div({
            "class": 'description',
            outlet: 'description'
          });
          return _this.div({
            "class": 'block'
          }, function() {
            _this.button({
              "class": 'inline-block update-instructions btn btn-success',
              outlet: 'updateButton'
            }, 'Restart and update');
            return _this.button({
              "class": 'inline-block btn',
              outlet: 'viewReleaseNotesButton'
            }, 'View past release notes');
          });
        };
      })(this));
    };

    ReleaseNotesView.prototype.getTitle = function() {
      return 'Release Notes';
    };

    ReleaseNotesView.prototype.getIconName = function() {
      return 'squirrel';
    };

    ReleaseNotesView.prototype.getUri = function() {
      return this.uri;
    };

    ReleaseNotesView.prototype.serialize = function() {
      return {
        deserializer: this.constructor.name,
        uri: this.uri,
        releaseNotes: this.releaseNotes,
        releaseVersion: this.releaseVersion
      };
    };

    ReleaseNotesView.prototype.initialize = function(uri, releaseVersion, releaseNotes) {
      this.uri = uri;
      this.releaseVersion = releaseVersion;
      this.releaseNotes = releaseNotes;
      if ((this.releaseNotes != null) && (this.releaseVersion != null)) {
        this.description.html(this.releaseNotes);
        this.version.text(this.releaseVersion);
        if (this.releaseVersion !== atom.getVersion()) {
          this.updateButton.show();
          this.subscribe(this.updateButton, 'click', function() {
            return atom.workspaceView.trigger('application:install-update');
          });
        }
      }
      return this.subscribe(this.viewReleaseNotesButton, 'click', function() {
        return shell.openExternal('https://atom.io/releases');
      });
    };

    return ReleaseNotesView;

  })(View);

}).call(this);
