// markdown-preview/lib/main.js
(function() {
  var MarkdownPreviewView, fs, renderer, url,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  url = require('url');

  fs = require('fs-plus');

  MarkdownPreviewView = require('./markdown-preview-view');

  renderer = null;

  module.exports = {
    configDefaults: {
      breakOnSingleNewline: false,
      grammars: ['source.gfm', 'source.litcoffee', 'text.plain', 'text.plain.null-grammar']
    },
    activate: function() {
      atom.workspaceView.command('markdown-preview:toggle', (function(_this) {
        return function() {
          return _this.toggle();
        };
      })(this));
      atom.workspaceView.command('markdown-preview:copy-html', (function(_this) {
        return function() {
          return _this.copyHtml();
        };
      })(this));
      atom.workspaceView.command('markdown-preview:toggle-break-on-single-newline', function() {
        return atom.config.toggle('markdown-preview.breakOnSingleNewline');
      });
      return atom.workspace.registerOpener(function(uriToOpen) {
        var error, host, pathname, protocol, _ref;
        try {
          _ref = url.parse(uriToOpen), protocol = _ref.protocol, host = _ref.host, pathname = _ref.pathname;
        } catch (_error) {
          error = _error;
          return;
        }
        if (protocol !== 'markdown-preview:') {
          return;
        }
        try {
          if (pathname) {
            pathname = decodeURI(pathname);
          }
        } catch (_error) {
          error = _error;
          return;
        }
        if (host === 'editor') {
          return new MarkdownPreviewView({
            editorId: pathname.substring(1)
          });
        } else {
          return new MarkdownPreviewView({
            filePath: pathname
          });
        }
      });
    },
    toggle: function() {
      var editor, grammars, previewPane, previousActivePane, uri, _ref, _ref1;
      editor = atom.workspace.getActiveEditor();
      if (editor == null) {
        return;
      }
      grammars = (_ref = atom.config.get('markdown-preview.grammars')) != null ? _ref : [];
      if (_ref1 = editor.getGrammar().scopeName, __indexOf.call(grammars, _ref1) < 0) {
        return;
      }
      uri = "markdown-preview://editor/" + editor.id;
      previewPane = atom.workspace.paneForUri(uri);
      if (previewPane) {
        previewPane.destroyItem(previewPane.itemForUri(uri));
        return;
      }
      previousActivePane = atom.workspace.getActivePane();
      return atom.workspace.open(uri, {
        split: 'right',
        searchAllPanes: true
      }).done(function(markdownPreviewView) {
        if (markdownPreviewView instanceof MarkdownPreviewView) {
          markdownPreviewView.renderMarkdown();
          return previousActivePane.activate();
        }
      });
    },
    copyHtml: function() {
      var editor, text;
      editor = atom.workspace.getActiveEditor();
      if (editor == null) {
        return;
      }
      if (renderer == null) {
        renderer = require('./renderer');
      }
      text = editor.getSelectedText() || editor.getText();
      return renderer.toText(text, editor.getPath(), (function(_this) {
        return function(error, html) {
          if (error) {
            return console.warn('Copying Markdown as HTML failed', error);
          } else {
            return atom.clipboard.write(html);
          }
        };
      })(this));
    }
  };

}).call(this);
