// settings-view/lib/available-package-view.js
(function() {
  var AvailablePackageView, View, shell, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  _ = require('underscore-plus');

  View = require('atom').View;

  shell = require('shell');

  module.exports = AvailablePackageView = (function(_super) {
    __extends(AvailablePackageView, _super);

    function AvailablePackageView() {
      return AvailablePackageView.__super__.constructor.apply(this, arguments);
    }

    AvailablePackageView.content = function(_arg) {
      var description, downloads, name;
      name = _arg.name, description = _arg.description, downloads = _arg.downloads;
      return this.div({
        "class": 'col-lg-4 available-package-view'
      }, (function(_this) {
        return function() {
          return _this.div({
            "class": 'thumbnail text'
          }, function() {
            return _this.div({
              "class": 'caption'
            }, function() {
              _this.span({
                outlet: 'status',
                "class": 'package-status icon'
              });
              _this.h4({
                "class": 'package-name native-key-bindings',
                tabindex: -1
              }, _.undasherize(_.uncamelcase(name)));
              if (downloads >= 0) {
                _this.p({
                  "class": 'downloads native-key-bindings',
                  tabindex: -1
                }, _.pluralize(downloads, 'download'));
              }
              _this.p({
                "class": 'description native-key-bindings',
                tabindex: -1
              }, description != null ? description : '');
              return _this.div({
                "class": 'btn-toolbar'
              }, function() {
                _this.button({
                  outlet: 'installButton',
                  "class": 'btn btn-primary'
                }, 'Install');
                _this.button({
                  outlet: 'learnMoreButton',
                  "class": 'btn btn-default'
                }, 'Learn More');
                return _this.button({
                  outlet: 'settingsButton',
                  "class": 'btn btn-default'
                }, 'Settings');
              });
            });
          });
        };
      })(this));
    };

    AvailablePackageView.prototype.initialize = function(pack, packageManager) {
      this.pack = pack;
      this.packageManager = packageManager;
      this.type = this.pack.theme ? 'theme' : 'package';
      this.handlePackageEvents();
      this.installButton.on('click', (function(_this) {
        return function() {
          if (_this.isInstalled()) {
            return _this.uninstall();
          } else {
            return _this.install();
          }
        };
      })(this));
      this.settingsButton.on('click', (function(_this) {
        return function() {
          var _ref;
          return (_ref = _this.parents('.settings-view').view()) != null ? _ref.showPanel(_this.pack.name) : void 0;
        };
      })(this));
      return this.learnMoreButton.on('click', (function(_this) {
        return function() {
          return shell.openExternal("https://atom.io/packages/" + _this.pack.name);
        };
      })(this));
    };

    AvailablePackageView.prototype.handlePackageEvents = function() {
      this.subscribeToPackageEvent('package-installed package-install-failed theme-installed theme-install-failed', (function(_this) {
        return function(pack, error) {
          _this.installButton.prop('disabled', false);
          if (error != null) {
            return _this.setStatusIcon('alert');
          } else {
            _this.setStatusIcon('check');
            _this.settingsButton.show();
            return _this.installButton.text('Uninstall');
          }
        };
      })(this));
      this.subscribeToPackageEvent('package-installing', (function(_this) {
        return function(pack) {
          _this.installButton.prop('disabled', true);
          _this.installButton.text('Install');
          return _this.setStatusIcon('cloud-download');
        };
      })(this));
      this.subscribeToPackageEvent('package-uninstalling', (function(_this) {
        return function(pack) {
          _this.installButton.prop('disabled', true);
          return _this.setStatusIcon();
        };
      })(this));
      this.subscribeToPackageEvent('package-uninstalled package-uninstall-failed theme-uninstalled theme-uninstall-failed', (function(_this) {
        return function(pack, error) {
          _this.installButton.prop('disabled', false);
          if (error != null) {
            return _this.setStatusIcon('alert');
          } else {
            _this.installButton.text('Install');
            _this.settingsButton.hide();
            return _this.setStatusIcon();
          }
        };
      })(this));
      if (this.isInstalled()) {
        this.installButton.text('Uninstall');
        return this.setStatusIcon('check');
      } else {
        this.settingsButton.hide();
        if (this.isDisabled()) {
          return this.installButton.prop('disabled', true);
        }
      }
    };

    AvailablePackageView.prototype.isInstalled = function() {
      return atom.packages.isPackageLoaded(this.pack.name);
    };

    AvailablePackageView.prototype.isDisabled = function() {
      return atom.packages.isPackageDisabled(this.pack.name);
    };

    AvailablePackageView.prototype.subscribeToPackageEvent = function(event, callback) {
      return this.subscribe(this.packageManager, event, (function(_this) {
        return function(pack, error) {
          if (pack.name === _this.pack.name) {
            return callback(pack, error);
          }
        };
      })(this));
    };

    AvailablePackageView.prototype.install = function() {
      this.packageManager.emit('package-installing', this.pack);
      return this.packageManager.install(this.pack, (function(_this) {
        return function(error) {
          var _ref;
          if (error != null) {
            return console.error("Installing " + _this.type + " " + _this.pack.name + " failed", (_ref = error.stack) != null ? _ref : error, error.stderr);
          }
        };
      })(this));
    };

    AvailablePackageView.prototype.uninstall = function() {
      this.packageManager.emit('package-uninstalling', this.pack);
      return this.packageManager.uninstall(this.pack, (function(_this) {
        return function(error) {
          var _ref;
          if (error != null) {
            return console.error("Uninstalling " + _this.type + " " + _this.pack.name + " failed", (_ref = error.stack) != null ? _ref : error, error.stderr);
          }
        };
      })(this));
    };

    AvailablePackageView.prototype.setStatusIcon = function(iconName) {
      this.status.removeClass('icon-check icon-alert icon-cloud-download');
      if (iconName) {
        this.status.addClass("icon-" + iconName);
      }
      this.status.destroyTooltip();
      switch (iconName) {
        case 'check':
          return this.status.setTooltip(_.capitalize("" + this.type + " installed"));
        case 'alert':
          return this.status.setTooltip(_.capitalize("" + this.type + " failed to install"));
        case 'cloud-download':
          return this.status.setTooltip(_.capitalize("" + this.type + " installing"));
      }
    };

    return AvailablePackageView;

  })(View);

}).call(this);
