// settings-view/lib/packages-panel.js
(function() {
  var $, $$, AvailablePackageView, ErrorView, PackageManager, PackageUpdateView, PackagesPanel, SettingEditorView, View, fs, path, _, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  path = require('path');

  _ = require('underscore-plus');

  fs = require('fs-plus');

  _ref = require('atom'), $ = _ref.$, $$ = _ref.$$, View = _ref.View;

  AvailablePackageView = require('./available-package-view');

  ErrorView = require('./error-view');

  PackageManager = require('./package-manager');

  PackageUpdateView = require('./package-update-view');

  SettingEditorView = require('./setting-editor-view');

  module.exports = PackagesPanel = (function(_super) {
    __extends(PackagesPanel, _super);

    function PackagesPanel() {
      return PackagesPanel.__super__.constructor.apply(this, arguments);
    }

    PackagesPanel.content = function() {
      return this.div((function(_this) {
        return function() {
          _this.div({
            "class": 'section packages'
          }, function() {
            _this.div({
              "class": 'section-heading icon icon-squirrel'
            }, function() {
              _this.span('Available Updates');
              return _this.button({
                outlet: 'updateAllButton',
                "class": 'pull-right btn btn-primary'
              }, 'Update All');
            });
            _this.div({
              outlet: 'updateErrors'
            });
            _this.div({
              outlet: 'checkingMessage',
              "class": 'alert alert-info featured-message icon icon-hourglass'
            }, 'Checking for updates\u2026');
            _this.div({
              outlet: 'noUpdatesMessage',
              "class": 'alert alert-info featured-message icon icon-heart'
            }, 'All of your installed packages are up to date!');
            return _this.div({
              outlet: 'updatesContainer',
              "class": 'container package-container'
            });
          });
          _this.div({
            "class": 'section packages'
          }, function() {
            _this.div({
              "class": 'section-heading icon icon-cloud-download'
            }, 'Install Packages');
            _this.div({
              "class": 'text padded native-key-bindings',
              tabindex: -1
            }, function() {
              _this.span({
                "class": 'icon icon-question'
              });
              _this.span('Packages are published to  ');
              _this.a({
                "class": 'link',
                outlet: "openAtomIo"
              }, "atom.io");
              return _this.span(" and are installed to " + (path.join(fs.getHomeDirectory(), '.atom', 'packages')));
            });
            _this.div({
              "class": 'editor-container padded'
            }, function() {
              return _this.subview('searchEditorView', new SettingEditorView());
            });
            _this.div({
              outlet: 'searchErrors'
            });
            _this.div({
              outlet: 'searchMessage',
              "class": 'alert alert-info search-message icon icon-search'
            });
            return _this.div({
              outlet: 'resultsContainer',
              "class": 'container package-container'
            });
          });
          return _this.div({
            "class": 'section packages'
          }, function() {
            _this.div({
              "class": 'section-heading icon icon-star'
            }, 'Featured Packages');
            _this.div({
              outlet: 'featuredErrors'
            });
            _this.div({
              outlet: 'loadingMessage',
              "class": 'alert alert-info featured-message icon icon-hourglass'
            }, 'Loading featured packages\u2026');
            _this.div({
              outlet: 'emptyMessage',
              "class": 'alert alert-info featured-message icon icon-heart'
            }, 'You have every featured package installed already!');
            return _this.div({
              outlet: 'featuredContainer',
              "class": 'container package-container'
            });
          });
        };
      })(this));
    };

    PackagesPanel.prototype.initialize = function(packageManager) {
      this.packageManager = packageManager;
      this.openAtomIo.on('click', (function(_this) {
        return function() {
          require('shell').openExternal('https://atom.io/packages');
          return false;
        };
      })(this));
      this.noUpdatesMessage.hide();
      this.searchMessage.hide();
      this.emptyMessage.hide();
      this.searchEditorView.setPlaceholderText('Search packages');
      this.searchEditorView.on('core:confirm', (function(_this) {
        return function() {
          var query;
          if (query = _this.searchEditorView.getText().trim()) {
            return _this.search(query);
          }
        };
      })(this));
      this.subscribe(this.packageManager, 'package-install-failed', (function(_this) {
        return function(pack, error) {
          return _this.searchErrors.append(new ErrorView(error));
        };
      })(this));
      this.subscribe(this.packageManager, 'package-update-failed theme-update-failed', (function(_this) {
        return function(pack, error) {
          return _this.updateErrors.append(new ErrorView(error));
        };
      })(this));
      this.updateAllButton.hide();
      this.updateAllButton.on('click', (function(_this) {
        return function() {
          var updateView, _i, _len, _ref1, _ref2, _results;
          _this.updateAllButton.prop('disabled', true);
          _ref1 = _this.updatesContainer.find('.available-package-view');
          _results = [];
          for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
            updateView = _ref1[_i];
            _results.push((_ref2 = $(updateView).view()) != null ? typeof _ref2.upgrade === "function" ? _ref2.upgrade() : void 0 : void 0);
          }
          return _results;
        };
      })(this));
      this.loadFeaturedPackages();
      return this.checkForUpdates();
    };

    PackagesPanel.prototype.focus = function() {
      return this.searchEditorView.focus();
    };

    PackagesPanel.prototype.search = function(query) {
      if (this.resultsContainer.children().length === 0) {
        this.searchMessage.text("Searching for \u201C" + query + "\u201D\u2026").show();
      }
      return this.packageManager.search(query, {
        packages: true
      }).then((function(_this) {
        return function(packages) {
          if (packages == null) {
            packages = [];
          }
          if (packages.length === 0) {
            _this.searchMessage.text("No package results for \u201C" + query + "\u201D").show();
          } else {
            _this.searchMessage.hide();
          }
          return _this.addPackageViews(_this.resultsContainer, packages);
        };
      })(this))["catch"]((function(_this) {
        return function(error) {
          _this.searchMessage.hide();
          return _this.searchErrors.append(new ErrorView(error));
        };
      })(this));
    };

    PackagesPanel.prototype.addPackageViews = function(container, packages) {
      var index, pack, packageRow, _i, _len, _results;
      container.empty();
      _results = [];
      for (index = _i = 0, _len = packages.length; _i < _len; index = ++_i) {
        pack = packages[index];
        if (index % 3 === 0) {
          packageRow = $$(function() {
            return this.div({
              "class": 'row'
            });
          });
          container.append(packageRow);
        }
        _results.push(packageRow.append(new AvailablePackageView(pack, this.packageManager)));
      }
      return _results;
    };

    PackagesPanel.prototype.addUpdateViews = function() {
      var index, pack, packageRow, _i, _len, _ref1, _results;
      if (this.availableUpdates.length > 0) {
        this.updateAllButton.show();
      }
      this.checkingMessage.hide();
      this.updatesContainer.empty();
      if (this.availableUpdates.length === 0) {
        this.noUpdatesMessage.show();
      }
      _ref1 = this.availableUpdates;
      _results = [];
      for (index = _i = 0, _len = _ref1.length; _i < _len; index = ++_i) {
        pack = _ref1[index];
        if (index % 3 === 0) {
          packageRow = $$(function() {
            return this.div({
              "class": 'row'
            });
          });
          this.updatesContainer.append(packageRow);
        }
        _results.push(packageRow.append(new PackageUpdateView(pack, this.packageManager)));
      }
      return _results;
    };

    PackagesPanel.prototype.filterPackages = function(packages) {
      return packages.filter(function(_arg) {
        var theme;
        theme = _arg.theme;
        return !theme;
      });
    };

    PackagesPanel.prototype.loadFeaturedPackages = function() {
      this.loadingMessage.show();
      return this.packageManager.getFeatured().then((function(_this) {
        return function(packages) {
          packages = _this.filterPackages(packages);
          _this.loadingMessage.hide();
          if (packages.length === 0) {
            _this.emptyMessage.show();
          }
          return _this.addPackageViews(_this.featuredContainer, packages);
        };
      })(this))["catch"]((function(_this) {
        return function(error) {
          _this.loadingMessage.hide();
          return _this.featuredErrors.append(new ErrorView(error));
        };
      })(this));
    };

    PackagesPanel.prototype.checkForUpdates = function() {
      this.checkingMessage.show();
      return this.packageManager.getOutdated().then((function(_this) {
        return function(availableUpdates) {
          _this.availableUpdates = availableUpdates;
          return _this.addUpdateViews();
        };
      })(this))["catch"]((function(_this) {
        return function(error) {
          _this.checkingMessage.hide();
          return _this.updateErrors.append(new ErrorView(error));
        };
      })(this));
    };

    return PackagesPanel;

  })(View);

}).call(this);
