// symbols-view/lib/file-view.js
(function() {
  var $$, FileView, SymbolsView, TagGenerator,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  $$ = require('atom').$$;

  SymbolsView = require('./symbols-view');

  TagGenerator = require('./tag-generator');

  module.exports = FileView = (function(_super) {
    __extends(FileView, _super);

    function FileView() {
      return FileView.__super__.constructor.apply(this, arguments);
    }

    FileView.prototype.initialize = function() {
      FileView.__super__.initialize.apply(this, arguments);
      this.cachedTags = {};
      this.subscribe(atom.project.eachBuffer((function(_this) {
        return function(buffer) {
          _this.subscribe(buffer, 'reloaded saved destroyed path-changed', function() {
            return delete _this.cachedTags[buffer.getPath()];
          });
          return _this.subscribe(buffer, 'destroyed', function() {
            return _this.unsubscribe(buffer);
          });
        };
      })(this)));
      return this.subscribe(atom.workspace.eachEditor((function(_this) {
        return function(editor) {
          _this.subscribe(editor, 'grammar-changed', function() {
            return delete _this.cachedTags[editor.getPath()];
          });
          return _this.subscribe(editor, 'destroyed', function() {
            return _this.unsubscribe(editor);
          });
        };
      })(this)));
    };

    FileView.prototype.viewForItem = function(_arg) {
      var name, position;
      position = _arg.position, name = _arg.name;
      return $$(function() {
        return this.li({
          "class": 'two-lines'
        }, (function(_this) {
          return function() {
            _this.div(name, {
              "class": 'primary-line'
            });
            return _this.div("Line " + (position.row + 1), {
              "class": 'secondary-line'
            });
          };
        })(this));
      });
    };

    FileView.prototype.toggle = function() {
      var filePath;
      if (this.hasParent()) {
        return this.cancel();
      } else if (filePath = this.getPath()) {
        this.populate(filePath);
        return this.attach();
      }
    };

    FileView.prototype.getPath = function() {
      var _ref;
      return (_ref = atom.workspace.getActiveEditor()) != null ? _ref.getPath() : void 0;
    };

    FileView.prototype.getScopeName = function() {
      var _ref, _ref1;
      return (_ref = atom.workspace.getActiveEditor()) != null ? (_ref1 = _ref.getGrammar()) != null ? _ref1.scopeName : void 0 : void 0;
    };

    FileView.prototype.populate = function(filePath) {
      var tags;
      this.list.empty();
      this.setLoading('Generating symbols\u2026');
      if (tags = this.cachedTags[filePath]) {
        this.maxItem = Infinity;
        return this.setItems(tags);
      } else {
        return this.generateTags(filePath);
      }
    };

    FileView.prototype.generateTags = function(filePath) {
      return new TagGenerator(filePath, this.getScopeName()).generate().done((function(_this) {
        return function(tags) {
          _this.cachedTags[filePath] = tags;
          _this.maxItem = Infinity;
          return _this.setItems(tags);
        };
      })(this));
    };

    return FileView;

  })(SymbolsView);

}).call(this);
